user/mgetty/voice/man/man3/wildmat.3
.\" Revision: 1.10 
.TH WILDMAT 3
.SH NAME
wildmat \- perform shell-style wildcard matching
.SH SYNOPSIS
.nf
.B "int"
.B "wildmat(text, pattern)"
.B "    char            *text;"
.B "    char            *pattern;"
.fi
.SH DESCRIPTION
.I Wildmat
compares the
.I text
against the
.I pattern
and
returns non-zero if the pattern matches the text.
The pattern is interpreted according to rules similar to shell filename
wildcards, and not as a full regular expression such as those handled by the
.IR grep (1)
family of programs or the
.IR regex (3)
or
.IR regexp (3)
set of routines.
.PP
The pattern is interpreted as follows:
.TP
.BI \e x
Turns off the special meaning of
.I x
and matches it directly; this is used mostly before a question mark or
asterisk, and is not special inside square brackets.
.TP
.B ?
Matches any single character.
.TP
.B *
Matches any sequence of zero or more characters.
.TP
.BI [ x...y ]
Matches any single character specified by the set
.IR x...y .
A minus sign may be used to indicate a range of characters.
That is,
.I [0\-5abc]
is a shorthand for
.IR [012345abc] .
More than one range may appear inside a character set;
.I [0-9a-zA-Z._]
matches almost all of the legal characters for a host name.
The close bracket,
.IR ] ,
may be used if it is the first character in the set.
The minus sign,
.IR \- ,
may be used if it is either the first or last character in the set.
.TP
.BI [^ x...y ]
This matches any character
.I not
in the set
.IR x...y ,
which is interpreted as described above.
For example,
.I [^]\-]
matches any character other than a close bracket or minus sign.
.SH HISTORY
Written by Rich $alz <rsalz@uunet.uu.net> in 1986, and posted to Usenet
several times since then, most notably in comp.sources.misc in
March, 1991.
.PP
Lars Mathiesen <thorinn@diku.dk> enhanced the multi-asterisk failure
mode in early 1991.
.PP
Rich and Lars increased the efficiency of star patterns and reposted it
to comp.sources.misc in April, 1991.
.PP
Robert Elz <kre@munnari.oz.au> added minus sign and close bracket handling
in June, 1991.
.PP
.de R$
This is revision \\$3, dated \\$4.
..
.R$ $Id: wildmat.3,v 1.1 1998/09/09 21:50:36 gert Exp $
.SH "SEE ALSO"
grep(1), regex(3), regexp(3).


freeswan/CHANGES
New and notable in 1.97:

Sanity check added for at least one of AH or ESP more for checking if
"make *config" has even been run.

Fixed 2.2 local IKE fragmentation blackhole.  Still won't work if
iptraf or another pcap app is running.

MCR added ikeping(8).

Applied DHR's tunnel patch to streamline IKE/specialSA processing.

Added local and remote IKE bypass tests to the UML test suite.

UML KLIPS %trap* tests now check for upbound pf_key messages.

An ICMP ICMP_DEST_UNREACH, ICMP_PKT_FILTERED is now sent upon %reject.

Added a UML input file with first and last packets being outside the 
covered range to "sandwich" the packets under test.

The default _updown script has been cleaned up slightly, reducing
internal duplication that arose accidentally.

When doing --add on an RSA-signature conn, ipsec_auto now tells Pluto
about the keys before, not after, telling it about the conn.  The main
consequence of this is that if something is wrong with one of the keys,
the conn itself does *not* get loaded.  This should reduce confusion.

The rp_filter checks during KLIPS startup have been fixed to be a bit less
paranoid; rp_filter is thought to cause problems only on the physical
interfaces, not on the ipsecN interfaces. 

The configuration process has been fixed to be a bit more paranoid.  Some
users have had mysterious compile failures with complaints involving names
like MD5_CTX; they seem to have resulted from kernel configuration changes
somehow getting into .config but not into include/linux/autoconf.h (which
is what actually controls most of the kernel compile).  The post-config
checking now verifies not only that .config has the right stuff in it, but
that autoconf.h does too.  (We still don't see *how* this discrepancy can
arise, but this will at least catch it when it does.)

The details of linking the KLIPS source into the kernel are now under
control of a configuration variable in Makefile.inc, to make it easier
for distribution builders to change.

The top-level Makefile now includes cleverness which assigns a date-based
version code if a build is done from a raw CVS checkout.

The library now includes a new function, ttodatav(), like ttodata() but
in some cases it gives more verbose error reports.  Also, there are now
a couple of functions which distill an RSA public key to a "key ID" for
use in messages, a crypto-quality pseudo-random-number generator, and a
function that converts a single address into a (singleton) subnet.

Barf's secret/key-censoring utilities have been tidied up so that what
they emit matches the key-ID library functions. 

The ipsec command now supplies a couple more environment variables to
things it calls, and this has reduced the number of things which need to
be edited at build time to know about configuration variables. 

The configuration reader now allows parameter names to begin with "_";
these are strictly reserved for internal development purposes and will
not in general be documented.


As usual, there are assorted small bug fixes and improvements to docs
and messages.



New and notable in 1.96:

Compatibility with the 2.0.xx kernels (under Red Hat 5.2) is broken.
This was accidental, and we have some hopes of fixing it, although it
won't last forever even if we do.  See README.

Our "IKE Implementation Issues" IETF draft has been added to our docs.

Newhostkey now has a mandatory --output "option" specifying a filename
(although the magic filename "-" means "standard output"), and sets
restrictive permissions on newly-created files to try to keep private keys
truly private. 

Pluto now gracefully handles a couple of odd error cases (missing private
keys, network errors at just the wrong time in negotiations) which used to
cause it to die, provoking an automatic restart of the IPsec subsystem. 

Makefile.inc now has more detailed control over where manpages go, to ease
integration into distributions.  The name of the destination directory for
RPMs is also now centralized there.

Rsasigkey now checks for certain obscure internal failures, which can
happen if --random is used to supply a source of random bits that aren't
really very random, and complains accordingly.

Barf's eroute table, formerly sorted by source, is now sorted by destination.
This is probably a bit clearer for hosts with lots of eroutes.

Use of the %reject target in a shunt eroute now, as long planned, actually
causes an ICMP rejection message (Destination Unreachable:  Communication
Administratively Prohibited). 

KLIPS now declares its module license to be "Dual BSD/GPL".  (It is 
dual because the libdes and zlib code is BSD licensed, while KLIPS itself 
is GPL).  Whether this is exactly right is not yet clear, but at least it
stops the latest module loaders from whining.

The UML test stuff continues to grow and add new test cases.

As usual, there are assorted small bug fixes and improvements to docs
and messages.



New and notable in 1.95:

The last remnants of the "%hold" bug, which broke 1.93 and 1.94, have
(we think) been dealt with.

Minor bugs in the RPM-building stuff have been fixed, and it has been
updated to deal with Red Hat's abrupt removal of the --buildroot option
from rpm.

Pluto now supports the ability to delete an instance of a conn, rather
than just the whole conn (for cases, notably OE, where a single connection
description can be involved in multiple connections).  There is no support
for this in ipsec_auto yet.

Pluto's routine (non-debug) logging is now somewhat terser, and some of
the more boring messages are produced only if --debug-lifecycle is turned
on (this is a temporary hack, it's not really debugging output).

Pluto's innards have been reorganized somewhat, in preparation for making
DNS lookups asynchronous.  There should be no user-visible effects.

Pluto now ignores IPv6 link-local interface addresses; we think that they
are never relevant, and trying to bind to them causes difficulties.  This

user/mgetty/voice/libutil/wildmat.c
/*
 * wildmat.c stolen from tin
 *
 * Added small modification to support a non C string for the pattern with
 * a given length.
 *
 * $Id: wildmat.c,v 1.1 1998/09/09 21:43:07 gert Exp $
 *
 */

/*
**  Revision: 1.9
**
**  Do shell-style pattern matching for ?, \, [], and * characters.
**  Might not be robust in face of malformed patterns; e.g., "foo[a-"
**  could cause a segmentation violation.  It is 8bit clean.
**
**  Written by Rich $alz, mirror!rs, Wed Nov 26 19:03:17 EST 1986.
**  Rich $alz is now <rsalz@osf.org>.
**  April, 1991:  Replaced mutually-recursive calls with in-line code
**  for the star character.
**
**  Special thanks to Lars Mathiesen <thorinn@diku.dk> for the ABORT code.
**  This can greatly speed up failing wildcard patterns.  For example:
**      pattern: -*-*-*-*-*-*-12-*-*-*-m-*-*-*
**      text 1:  -adobe-courier-bold-o-normal--12-120-75-75-m-70-iso8859-1
**      text 2:  -adobe-courier-bold-o-normal--12-120-75-75-X-70-iso8859-1
**  Text 1 matches with 51 calls, while text 2 fails with 54 calls.  Without
**  the ABORT code, it takes 22310 calls to fail.  Ugh.  The following
**  explanation is from Lars:
**  The precondition that must be fulfilled is that DoMatch will consume
**  at least one character in text.  This is true if *p is neither '*' nor
**  '\0'.)  The last return has ABORT instead of FALSE to avoid quadratic
**  behaviour in cases like pattern "*a*b*c*d" with text "abcxxxxx".  With
**  FALSE, each star-loop has to run to the end of the text; with ABORT
**  only the last one does.
**
**  Once the control of one instance of DoMatch enters the star-loop, that
**  instance will return either TRUE or ABORT, and any calling instance
**  will therefore return immediately after (without calling recursively
**  again).  In effect, only one star-loop is ever active.  It would be
**  possible to modify the code to maintain this context explicitly,
**  eliminating all recursive calls at the cost of some complication and
**  loss of clarity (and the ABORT stuff seems to be unclear enough by
**  itself).  I think it would be unwise to try to get this into a
**  released version unless you have a good test data base to try it out
**  on.
*/

#if __STDC__
static int DoMatch (char *text, char *p, int length);
#else
static int DoMatch (/*char *text, char *p, int length*/);
#endif

#define TRUE                    1
#define FALSE                   0
#define ABORT                   -1



    /* What character marks an inverted character class? */
#define NEGATE_CLASS            '^'
    /* Is "*" a common pattern? */
#define OPTIMIZE_JUST_STAR
    /* Do tar(1) matching rules, which ignore a trailing slash? */
#undef MATCH_TAR_PATTERN


/*
**  Match text and p, return TRUE, FALSE, or ABORT.
*/

static int
DoMatch(text, p, length)
    register char       *text;
    register char       *p;
    register int        length;
{
#ifndef INDEX_DAEMON

    register int        last;
    register int        matched;
    register int        reverse;
    register char       *start;

    for (start = p ; (p - start) < length; text++, p++) {
        if (*text == '\0' && *p != '*')
            return ABORT;
        switch (*p) {
        case '\\':
            /* Literal match with following character. */
            p++;
            /* FALLTHROUGH */
        default:
            if (*text != *p)
                return FALSE;
            continue;
        case '?':
            /* Match anything. */
            continue;
        case '*':
            while (*++p == '*')
                /* Consecutive stars act just like one. */
                continue;
            if ((p - start) == length)
                /* Trailing star matches everything. */
                return TRUE;
            while (*text)
                if ((matched = DoMatch(text++, p, length + start - p)) != FALSE)
                    return matched;
            return ABORT;
        case '[':
            reverse = p[1] == NEGATE_CLASS ? TRUE : FALSE;
            if (reverse)
                /* Inverted character class. */
                p++;
            matched = FALSE;
            if (p[1] == ']' || p[1] == '-')
                if (*++p == *text)
                    matched = TRUE;
            for (last = *p; *++p && *p != ']'; last = *p)
                /* This next line requires a good C compiler. */
                if (*p == '-' && p[1] != ']'
                    ? *text <= *++p && *text >= last : *text == *p)
                    matched = TRUE;
            if (matched == reverse)
                return FALSE;
            continue;
        }
    }

#ifdef  MATCH_TAR_PATTERN
    if (*text == '/')
        return TRUE;
#endif  /* MATCH_TAR_ATTERN */

#endif  /* INDEX_DAEMON */

    return *text == '\0';
}


/*
**  User-level routine.  Returns TRUE or FALSE.
*/
#if __STDC__
int wildmat(char *text, char *p, int length)
#else
int
wildmat(text, p, length)
    char        *text;
    char        *p;
    int         length;
#endif
{
#ifdef  OPTIMIZE_JUST_STAR
    if (p[0] == '*' && (length == 1))
        return TRUE;
#endif  /* OPTIMIZE_JUST_STAR */
    return DoMatch(text, p, length) == TRUE;
}
